app/scripts/lib/transaction-metrics.ts

```diff
@@ -0,0 +1,879 @@
+        category: MetaMetricsEventCategory.Transactions,
+        successEvent: TransactionMetaMetricsEvent.finalized,
+        properties: payload.properties,
+        sensitiveProperties: payload.sensitiveProperties,
+        actionId: transactionMeta.actionId,
+        uniqueIdentifier,
+        persist: true,
+      });
+      break;
+    default:
+      break;
+  }
+}
+
+function updateTransactionEventFragment({
+  eventName,
+  controllerActions,
+  transactionMeta,
+  payload,
+}: {
+  eventName: TransactionMetaMetricsEvent;
+  controllerActions: ControllerActions;
+  transactionMeta: TransactionMeta;
+  payload: any;
+}) {
+  const uniqueId = getUniqueId(eventName, transactionMeta.id);
+
+  switch (eventName) {
+    case TransactionMetaMetricsEvent.approved:
+      controllerActions.updateEventFragment(uniqueId, {
+        properties: payload.properties,
+        sensitiveProperties: payload.sensitiveProperties,
+      });
+      break;
+
+    case TransactionMetaMetricsEvent.rejected:
+      controllerActions.updateEventFragment(uniqueId, {
+        properties: payload.properties,
+        sensitiveProperties: payload.sensitiveProperties,
+      });
+      break;
+
+    case TransactionMetaMetricsEvent.finalized:
+      controllerActions.updateEventFragment(uniqueId, {
+        properties: payload.properties,
+        sensitiveProperties: payload.sensitiveProperties,
+      });
+      break;
+    default:
+      break;
+  }
+}
+
+function finalizeTransactionEventFragment({
+  eventName,
+  controllerActions,
+  transactionMeta,
+}: {
+  eventName: TransactionMetaMetricsEvent;
+  controllerActions: ControllerActions;
+  transactionMeta: TransactionMeta;
+}) {
+  const uniqueId = getUniqueId(eventName, transactionMeta.id);
+
+  switch (eventName) {
+    case TransactionMetaMetricsEvent.approved:
+      controllerActions.finalizeEventFragment(uniqueId);
+      break;
+
+    case TransactionMetaMetricsEvent.rejected:
+      controllerActions.finalizeEventFragment(uniqueId, {
+        abandoned: true,
+      });
+      break;
+
+    case TransactionMetaMetricsEvent.finalized:
+      controllerActions.finalizeEventFragment(uniqueId);
+      break;
+    default:
+      break;
+  }
+}
+
+function hasFragment(
+  getEventFragmentById: (arg0: string) => any,
+  eventName: TransactionMetaMetricsEvent,
+  transactionMeta: TransactionMeta,
+) {
+  const uniqueId = getUniqueId(eventName, transactionMeta.id);
+  const fragment = getEventFragmentById(uniqueId);
+  return typeof fragment !== 'undefined';
+}
+
+function getUniqueId(
+  eventName: TransactionMetaMetricsEvent,
+  transactionId: string,
+) {
+  const isSubmitted = [
+    TransactionMetaMetricsEvent.finalized,
+    TransactionMetaMetricsEvent.submitted,
+  ].includes(eventName);
+  const uniqueIdentifier = `transaction-${
+    isSubmitted ? 'submitted' : 'added'
+  }-${transactionId}`;
+
+  return uniqueIdentifier;
+}
+
+async function buildEventFragmentProperties({
```

Comment:
Not sure how far we want to get to cover all branches in this function. I leave it for later but feel free to comment